### ISSUE TITLE
[no-unnecessary-condition] false positive if local variable modified in forEach

### ISSUE THREAD

**Issue by robstoll** [typescript-eslint/typescript-eslint#4624]
- [x] I have tried restarting my IDE and the issue persists.
- [ ] I have updated to the latest version of the packages.
- [ ] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**
```JSON
{
  "rules": {
     "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
let c = false;
const a = [1, 2, 3];
a.forEach(_ => {
  c = true;
});
if (c) { // <- no-unnecessary-condition
}
for (const _ of a) {
  c = true;
}
if (c) { // works as expected, no complaint
}
```
**Expected Result**

also changes in a function should be taken into account during flow-analysis

**Actual Result**
false positive

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.2` |
| `@typescript-eslint/parser`        | `5.10.2` |
| `TypeScript`                       | `~4.5.5` |
| `ESLint`                           | `^8.2.0` |
| `node`                             | `>=16.13.2` |

**Comment by JoshuaKGoldberg** [comment]
Unfortunately, there's nothing we can do here. Your issue is really with TypeScript not being able to tell that `.forEach` runs synchronously. https://github.com/microsoft/TypeScript/issues/45181 -> https://github.com/microsoft/TypeScript/issues/9998

If you copy & paste that code snippet into typescriptlang.org.play and check the hover information for the `c` in the first `if (c) {`, you'll see that TypeScript infers its type to be `false`.

**Comment by robstoll** [comment]
Fair enough, thanks for the information
